include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,347 @@
+            }
+        );
+    }
+
+    template <typename T>
+    array_data make_array_data_for_dictionary_encoded_layout()
+    {
+        return {
+            .type = data_descriptor(arrow_type_id<std::uint64_t>()),
+            .length = 0,
+            .offset = 0,
+            .bitmap = {},
+            .buffers = {array_data::buffer_type(sizeof(std::int64_t), 0)},
+            .child_data = {},
+            .dictionary = value_ptr<array_data>(make_array_data_for_variable_size_binary_layout<T>())
+        };
+    }
+
+    template <constant_range_for_array_data ValueRange>
+    array_data make_array_data_for_dictionary_encoded_layout(
+        ValueRange&& values,
+        const array_data::bitmap_type& bitmap,
+        std::int64_t offset
+    )
+    {
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const values_and_indexes<const std::ranges::range_value_t<ValueRange>> vec_and_indexes{values};
+        const auto& indexes = vec_and_indexes.indexes;
+        const auto create_buffer = [&indexes]()
+        {
+            const size_t buffer_size = indexes.size() * sizeof(size_t) / sizeof(uint8_t);
+            array_data::buffer_type b(buffer_size);
+            std::ranges::copy(indexes, b.data<size_t>());
+            return b;
+        };
+        return {
+            .type = data_descriptor(arrow_type_id<std::uint64_t>()),
+            .length = static_cast<array_data::length_type>(indexes.size()),
+            .offset = offset,
+            .bitmap = bitmap,
+            .buffers = {create_buffer()},
+            .child_data = {},
+            .dictionary = value_ptr<array_data>(make_array_data_for_variable_size_binary_layout(
+                vec_and_indexes.values,
+                array_data::bitmap_type(vec_and_indexes.values.size(), true),
+                0
+            ))
+        };
+    }
+
+    template <is_a_supported_layout Layout>
+    array_data make_default_array_data_factory()
+    {
+        if constexpr (mpl::is_type_instance_of_v<Layout, fixed_size_layout>)
+        {
+            return make_array_data_for_fixed_size_layout<typename Layout::inner_value_type>();
+        }
+        else if constexpr (mpl::is_type_instance_of_v<Layout, variable_size_binary_layout>)
+        {
+            return make_array_data_for_variable_size_binary_layout<typename Layout::inner_value_type>();
+        }
+        else if constexpr (mpl::is_type_instance_of_v<Layout, dictionary_encoded_layout>)
+        {
+            return make_array_data_for_dictionary_encoded_layout<typename Layout::inner_value_type>();
+        }
+        else
+        {
+            static_assert(
+                mpl::dependent_false<Layout>::value,
+                "Unsupported layout type. Please check the is_a_supported_layout concept and create a function make_array_data_for_... for the new layout type."
+            );
+        }
```

Comment:
Suspicion: this will not work on some compilers because we dont return anything in that branch. I dont know if `std::unreachable()` (C++23, but implementable) works in that case... 
Anyway if no compiler in the CI have trouble with this, ignore this comment.

Reply:
Done